[1-3단계 - 방탈출 예약 관리] 멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
[멍구 프롤로그 (1단계만 작성)](https://prolog.techcourse.co.kr/studylogs/4166)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
List<Reservation>와 현재 Index를 관리하는 역할을 Repository로 분리함으로써 Controller의 단일 책임이 잘 지켜진 것 같습니다. 다만 요구사항을 구현하는 데 있어서 모든 원시 값과 문자열을 포장한다 등의 생활체조 원칙을 지키지는 않았습니다. 현재 요구사항 수준에서는 감쌀 필요가 없다고 생각했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

간단한 게시판 예제를 클론코딩 해본 경험이 있습니다(배포 X). 스프링으로 프로젝트 경험은 없습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

전체적으로 Spring을 올바르게 사용하였는지 궁금합니다. 스프링 경험이 거의 없어서, 어떤 부분이 중요한지 잘 모르는 수준입니다. 자유롭게 리뷰해주시면 감사하겠습니다!!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

yuyoungrhee committed:
docs(README): 요구사항 정리

yuyoungrhee committed:
feat(controller): 1단계 요구사항 구현

- admin 페이지가 index.html을 보여주도록 구현

yuyoungrhee committed:
feat(domain): Reservation 정의

yuyoungrhee committed:
feat: 예약 조회 페이지 구현

yuyoungrhee committed:
feat: 예약 json 조회 구현

yuyoungrhee committed:
fix: 서버 사이드 렌더링 버그 수정

- 타임리프 기능 삭제
- 컨트롤러 모델을 통한 화면 구성에서 json을 통한 화면 구성으로 변경

yuyoungrhee committed:
feat: 예약 추가 api 구현

yuyoungrhee committed:
feat: 예약 삭제 api 구현

yuyoungrhee committed:
refactor: 저장 로직을 Repository로 분리

yuyoungrhee committed:
test(repository): 예약Repository 테스트 추가

yuyoungrhee committed:
refactor(dto): 도메인 변환 로직 이동

yuyoungrhee committed:
docs(README): 구현한 기능 목록 업데이트

yuyoungrhee commented on src/main/java/roomescape/dto/request/ReservationRequest.java:
```diff
@@ -0,0 +1,20 @@
+        String name,
+        String time
+) {
+    public Reservation dtoToReservationWithoutId(){
+        LocalDate date = LocalDate.parse(date());
+        LocalTime time = LocalTime.parse(time());
+
+        LocalDateTime reservationTime = LocalDateTime.of(date, time);
+        return Reservation.createReservationWithoutID(name(), reservationTime);
+    }
```
Dto 내부에 dto를 객체로 변환하는 로직이 존재하는데, dto의 책임으로 보아도 괜찮을까요? 
현재는 Servie가 따로 존재하지 않는 구조이기에 Controller, Dto 둘 중 하나를 선택해야한다면 Dto에 두는 게 맞다고 생각하여 dto 내부에 두었습니다. Controller는 흐름만 담당해야한다고 생각했고, 비즈니스 로직을 최대한 덜어내는 쪽으로 구현했습니다.

dusdn1702 replied:
선택지가 Controller, DTO 둘 중 하나였던 이유가 있을까요?
request를 domain으로 변환하는 로직으로 보이는데, 여기서 request에 대한 유효성 검증이 필요하다면 어디서 검증해야한다고 생각하시나요?
dto의 책임은 무엇이라고 생각하셨는지도 궁금해요!

yuyoungrhee replied:
### dto <-> domain 변환 로직
Service 계층이 이 변환 로직을 가지는 게 가장 좋다고 생각했습니다. controller는 순수하게 흐름만 제어, dto는 단순한 데이터만 담당한다면 변환 로직은 자연스럽게 양측 계층을 모두 알고있는 service에서 담당하게 됩니다.

그런데 현재 코드에서 변환을 위해 Service계층을 도입하는 건 복잡도가 올라가고 이점이 없다고 느껴졌습니다. 따라서 선택지가 Controller, Dto 둘 중 하나였고 Dto를 선택했습니다.

**Dto 내부 변환 문제점 -> Mapepr 도입**
Dto가 Domain을 알아야해서 Domain에 영향을 받거나 주게 됩니다. 따라서 Dto 내부도 적합하지 않다고 생각했고, **별도의 Mapper 클래스를 도입**하게 되었습니다. 이런 구조가 괜찮을까요?

[refactor(mapper): dto와 domain의 변환을 담당하는 ReservationMapper 구현](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/68e56218a3e06179c4a8aa15681b2961ea190d3a)

---
### Dto 책임은 무엇이라고 생각하는지
dto는 계층 간 데이터 전달을 위한 객체입니다. dto의 책임에는 데이터 보관, 도메인 보호, JSON 응답 or 요청 구조 제어 등이 있습니다. 

### Request -> Domain 변환 시 유효성 검증 위치
검증의 성격에 따라 위치를 분리할 것 같습니다.

Request(Dto): 입력 값 자체의 형식/제약 조건 검증
Domain: 비즈니스 규칙 기반의 검증

예를 들어 
- date가 null 또는 `20212-01-01`처럼 잘못된 요청이 온다면 Dto에서 검증합니다.
- 예약 인원이 1명 이상,8명 이하라던가 등의 규칙은 도메인의 성질 자체, 변하지 않는 도메인 규칙이므로 Reservation에서 검증합니다.

dusdn1702 replied:
1. dto <-> domain 변환 로직
현재 구조에서 controller에선 흐름만 제어하도록 했기 때문에 mapper를 도입하는 것도 방안 중에 하나라고 보여지네요. 좋습니다 😄 
2. Request -> Domain 변환 시 유효성 검증 위치
좋습니다 👍

yuyoungrhee committed:
refactor: 컨벤션에 맞도록 코드 수정

dusdn1702 commented on README.md:
```diff
@@ -0,0 +1,16 @@
+# 방탈출 예약 관리 1-3단계
+
+## 🚀 1단계 - 홈 화면
+- [x] localhost:8080/admin 요청 시 아래 화면과 같이 어드민 메인 페이지가 응답할 수 있도록 구현하세요.
```
현재 localhost:8080 요청 시에 whitelabel 에러로 응답되는 부분 개선해보면 좋을 것 같아요.
`스프링부트 프로젝트에서 welcome page를 응답하는 방법을 참고해보세요.`

yuyoungrhee replied:
개선했습니다! 아래쪽에 있는 요구사항이라 놓친 것 같습니다.

index.html을 반환하는 방법과 /admin/reservation 응답 페이지를 반환하는 방법 중에 고민했는데, 현재 예약 페이지가 홈화면의 기능을 한다고 생각해 /admin/reservation으로 redirect 하도록 구현했습니다.

[feat(controller): localhost:8080 요청 시에도 예약 페이지가 보여지도록 구현 추가](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/45833af0b7c50e1b445e78970c10371fa582f57e)

dusdn1702 commented on src/main/java/roomescape/controller/RoomEscapeController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity<Void> deleteReservation(@PathVariable("id") Long id) {
```
ResponseEntity 를 사용하여 응답한 이유가 궁금해요!

yuyoungrhee replied:
학습테스트를 그대로 적용하다가 코드 일관성을 잃어버린 것 같습니다..😅

deleteReservation이 `204 No Content `등 다른 상태를 반환해야했다면 `ResponseEntity`를 사용할 이유가 충분하다고 생각하는데, 현재는 `200 OK` 를 반환하는 것이 요구사항이므로 `@ResponseBody + void 반환` 으로 처리해도 충분할 것 같습니다. 코드도 더 간결합니다.

[refactor(controller): deleteReservation() 메서드도](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/d153b61f718de469c327e548c72cbced1163206f) `@Responsebody`[를 통해 응답을 반환하도록 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/d153b61f718de469c327e548c72cbced1163206f)](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/d153b61f718de469c327e548c72cbced1163206f)

dusdn1702 commented on src/main/java/roomescape/controller/RoomEscapeController.java:
```diff
@@ -0,0 +1,49 @@
+    @ResponseBody
+    @GetMapping("/reservations")
+    public List<ReservationResponse> reservationsJson() {
+        return ReservationResponse.reservationsToDtos(repository.findAll());
```
repository에서 모든 예약들을 가져오는 로직과 변환 로직을 분리하면 좋겠어요.
```suggestion
        List<Reservations> reservations = repository.findAll();
        return ReservationResponse.reservationsToDtos(reservations);
```

yuyoungrhee replied:
반영했습니다! 중간 변수를 제거하는 것이 더 간결하다고 생각했는데, 코드의 명확성과 이해를 위해서는 분리하는 것이 더 유리하다고 생각합니다.

[refactor(controller): findAll() 호출 결과를 변수로 추출하여 가독성 향상](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/27b47cb959402da599eaae513b2353fabe877507)

dusdn1702 commented on src/main/java/roomescape/controller/RoomEscapeController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @ResponseBody
+    @PostMapping("/reservations")
+    public ReservationResponse addReservation(@RequestBody ReservationRequest request) {
```
예약 추가 요청 응답에 ReservationResponse를 주신 이유가 무엇인가요??

yuyoungrhee replied:
예약 추가 API Response 명세는 다음과 같습니다

```http
HTTP/1.1 200 
Content-Type: application/json

{
    "id": 1,
    "name": "브라운",
    "date": "2023-08-05",
    "time": "15:40"
}
```

예약 추가 응답으로 Reservation 객체를 그대로 주게 되면 아래와 같은 Http 응답을 보내게 됩니다.
``` http
{
    "id": 1,
    "name": "브라운",
    "reservationTime": "2023-08-05T18:00:00"
}
```

원하는 응답 형태로 보내기 위해서는 여러가지 방법이 있으나 Dto 사용이 가장 깔끔하고 JSON 변환도 자동으로 이루어지기 때문에 Response형식 Dto를 만드는 방법으로 해결했습니다.

다른 방법 1. @JsonProperty + @JsonIgnore 로 Reservation의 date, time을 나눠서 @JsponProperty로 등록, 기존 reservationTime은 @JsonIgnore로 무시
다른 방법 2. 커스텀 Serializer 등록 및 사용
다른 방법 3. Map<String, String> 형태를 만들어서 반환

dusdn1702 replied:
다른 방법들과 현재의 방식 채택 이유까지 좋네요 👍

dusdn1702 commented on src/main/java/roomescape/dto/response/ReservationResponse.java:
```diff
@@ -0,0 +1,31 @@
+
+    public static ReservationResponse reservationToDto(Reservation reservation) {
+        LocalDateTime dateTime = reservation.getReservationTime();
+        DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
```
DateFormatter는 상수로 관리하면 좋을 것 같아요.

yuyoungrhee replied:
동의합니다! 반영했습니다.

[refactor(ReservationResponse): DateTimeFormatter를 상수로 관리](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/9e18b2fd965a8f9d61ba27696df3c3fc896cb835)

dusdn1702 commented on src/main/java/roomescape/dto/response/ReservationResponse.java:
```diff
@@ -0,0 +1,31 @@
+        String time
+) {
+
+    public static ReservationResponse reservationToDto(Reservation reservation) {
```
static을 사용하신 기준이 무엇인가요?

yuyoungrhee replied:
static 사용O: 객체 상태를 사용하지 않는 경우 / 순수 유틸 or 공통 헬퍼 메서드
static 사용 X: 객체의 상태를 활용해야하는 경우, 도메인의 비즈니스 로직

Dto는 도메인이 아니고, 중요한 비즈니스로직이라기 보다는 유틸느낌, 그리고 정확히는 정적팩터리메서드의 역할(자기 자신을 생성)이기 때문에  static을 사용했습니다.

반대로 ReservationRequest의 Reservation으로의 변환로직은 내부 필드 상태(this)를 활용하고, Request 자신이 가지고 있는 데이터를 변환하는 역할이기 때문에 static을 사용하지 않았습니다.

dusdn1702 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.time.LocalDateTime;
+
+public class Reservation {
```
Reservation은 record로 관리하지 않은 이유가 궁금해요!

yuyoungrhee replied:
record는 데이터를 담기 위한 클래스를 작성할 때 보일러플레이트의 부담을 줄이기 위해서 도입되었습니다. 때문에 record의 사용목적은 데이터를 담기 위한 클래스, 불변, 메서드 자동 생성 이어야 한다고 생각합니다.

이번에 공식문서 관련해서 OT를 들어서 record의 공식문서를 읽어보았는데 record는 데이터를 담기 위한 용기 역할이다 라는 문장이 있습니다.

> “Records are a transparent carrier for immutable data.”

Reservation을 record로 관리하게 된다면 예약 데이터 전달 책임 + 예약에 대한 전반적인 비즈니스 행위 책임이 생기게 되어서 SRP를 위반하게 됩니다. 현재 Reservation은 비즈니스 로직이 존재하지 않지만 **예약 시스템에서 핵심적인 도메인이라고 생각해서** 확장이나 기능 추가 가능성이 있다고 생각했습니다.

dusdn1702 replied:
좋습니다 ㅎㅎ 👍 💯

dusdn1702 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,47 @@
+
+    private final List<Reservation> reservations = new ArrayList<>();
+
+    private AtomicLong index = new AtomicLong(3);
```
AtomicLong 👍

yuyoungrhee replied:
요구사항 힌트에 있었던 내용이라 AtomicLong을 사용했습니다!

더 찾아보니 스프링은 기본적으로 멀티 스레드 환경이기 때문에 동시성 문제를 해결하고 안전하게 id를 증가시키기 위해 AtomicLong을 써야합니다. 
- 동시에 addReservation을 요청한다면 id 중복이 생길 수도 있습니다. 
- incrementAndGet() 메서드는 내부적으로 Compare And Swap을 사용해서 동시 접근을 안전하게 처리합니다. 

그런데 관련해서 공부하면서 ArrayList도 멀티 스레드 환경에서 동시성 문제가 있다는 것을 알았습니다. ArrayList는 synchronizedList로 해결했습니다. 
[fix(ReservationRepository): List도 동시성에 안전하도록 synchronizedList 사용하도록 수정](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/7940974f5ca9d3e2773da8f577c605235845cf57)

**List 동시성 문제를 해결하는 방법**
1. synchronizedList
    - 내부적으로 모든 메서드에 synchronized 걸려있음 -> 안전 / 성능 저하
    - 간단함
2. CopyOnWriteArrayList
    - 쓰기 성능 느림
    - add나 remove 할 때마다 배열 전체 복사: 예약 개수가 많아질수록 성능 매우 저하, 메모리 낭비
    - 읽기 성능 좋음: 읽을 때는 락 없이 읽음

현재로서는 둘 중 어느 것으로 해결해도 상관 없다고 생각했습니다만 상대적으로 쓰기 / 삭제도 많이 발생하는 애플리케이션 구조인 것 같아서 synchronizedList로 해결해보았습니다.

이 정도로 찾아봤는데, 혹시 빼먹거나 더 공부해야할 부분이 있다면 알려주세요..!

dusdn1702 replied:
사용이 좋아서 코멘트 달았어요. 추가로 학습까지 고생 많으셨습니다.
다만, 여기서 넣는 3이라는 값과 디폴트로 추가해주는 데이터의 갯수를 일치시키는게 좋을 것 같은데, 어떠신가요?

dusdn1702 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,47 @@
+
+    public ReservationRepository() {
+        List<Reservation> initialReservations = List.of(
+                Reservation.createReservation(1L, "브라운", LocalDateTime.of(2024, 4, 1, 10, 0)),
```
최초 데이터 생성 시 createReservationWithId가 아닌 다른 메서드를 사용하신 이유가 무엇일까요??

yuyoungrhee replied:
`createReservation`은 `Reservation`에 필요한 모든 값을 직접 표시하는 메서드입니다. 따라서 id도 지정해줄 수 있습니다. 

Resository에 최초 데이터를 삽입할 때 createReservation으로 명시적으로 입력하는 게 더 직관적이고 간단하다고 생각해서 사용했습니다.

createReservationWithId()를 사용한다면,
``` java
Reservation.createReservationWithId(1L, new Reservation(null, "브라운", ...))
```

이렇게 한 단계를 더 거쳐야되기 때문에 가독성이 낮아지는 것 같습니다.

dusdn1702 replied:
이 상황에서 저는 말씀 주신 것처럼 사용을 해도 괜찮다고 생각했어요. 
명확한 id 확인이 필요하기 때문이고, 3개의 객체 추가 사용은 문제가 되지 않는다고 보았거든요.
멍구가 해당 이유로 인해서 다른 메서드를 사용하신 부분도 이해되어 좋습니다 👍

dusdn1702 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public Long add(Reservation reservation) {
+        long currentIndex = index.incrementAndGet();
```
👍👍👍

dusdn1702 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,47 @@
+        return reservations.stream()
+                .filter(reservation -> reservation.sameId(id))
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("해당 ID가 존재하지 않습니다."));
```
에러 처리까지 좋네요!

dusdn1702 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,20 +1,97 @@
+                .when().get("/reservations")
+                .then().log().all()
+                .statusCode(200)
+                .body("size()", is(3)); // 아직 생성 요청이 없으니 Controller에서 임의로 넣어준 Reservation 갯수 만큼 검증하거나 0개임을 확인하세요.
```
주석에 적혀있는 것처럼 repository에서 3개의 최초 데이터를 생성하고 있는데, 이 갯수가 변경되면 이 값도 변경이 필요하겠네요. 같은 값으로 관리해줄 수 있는 방법이 있을까요?

yuyoungrhee replied:
repository의 최초 데이터 개수를 직접 계산해서 테스트하도록 변경했습니다!

[test: 테스트 추가 및 리팩터링](https://github.com/woowacourse/spring-roomescape-admin/pull/200/commits/e6c89e00d76ad2080df216c840e8a43522eb04a7)

dusdn1702 commented on src/test/java/roomescape/repository/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,90 @@
+        assertThat(reservation.getReservationTime()).isEqualTo(LocalDateTime.of(2024, 4, 1, 10, 0));
+    }
+
+    @DisplayName("없는 id로 찾으려고 할 때 예외가 발생한다.")
```
예외 테스트까지 💯

dusdn1702 reviewed (changes requested):
안녕하세요 멍구~
레벨 2 첫 미션 수행하시느라 고생 많으셨습니다.
스프링 사용 경험이 적음에도 잘 구현해주셨네요 👍 
몇가지 코멘트 남겨두었습니다. 확인 부탁드려요!

yuyoungrhee committed:
refactor(controller): findAll() 호출 결과를 변수로 추출하여 가독성 향상

yuyoungrhee committed:
refactor(controller): deleteReservation() 메서드도 @ResponseBody를 통해 응답을 반환하도록 변경

yuyoungrhee committed:
feat(controller): localhost:8080 요청 시에도 예약 페이지가 보여지도록 구현 추가

yuyoungrhee committed:
refactor(ReservationResponse): DateTimeFormatter를 상수로 관리

yuyoungrhee committed:
refactor: 컨트롤러에 Repository의 의존성주입(DI)

- controller 내부에서 new Repository -> @Component 등록된 Repository + controller에서 Repository 생성자주입 DI로 변경

yuyoungrhee committed:
test: 테스트 추가 및 리팩터링

- 테스트 이름 변경
- 숫자를 지정하는 대신 repository의 실제 데이터 개수를 통해 테스트하도록 변경

yuyoungrhee committed:
fix(ReservationRepository): List도 동시성에 안전하도록 synchronizedList 사용하도록 수정

yuyoungrhee committed:
refactor(controller): 뷰 반환용, API 응답용 컨트롤러를 분리

기본 페이지 반환, admin 페이지 반환, reservations 관련 api 응답 으로 컨트롤러 분리

yuyoungrhee committed:
refactor(mapper): dto와 domain의 변환을 담당하는 ReservationMapper 구현

- 기존 dto에서 변환 담당 로직을 ReservationMapper로 이동

yuyoungrhee commented:
## 구조적 변경 사항
큰 변경사항만 정리했습니다! 나머지는 코멘트에 답변 달아두었습니다.

### Controller 분리
- view 반환 로직과 API 응답(JSON) 로직이 섞여있는 구조이기 때문에 분리를 해야겠다고 느꼈습니다. 

1. Home Controller: localhost:8080 으로 요청할 시 admin/reservation (예약 페이지)로 리다이렉션합니다.
2. AdminController: HTML을 반환합니다.
3. ReservationController: JSON을 반환합니다. @ResonseBody를 붙여서 뷰가 아니라 JSON을 반환해야하기 때문에 `@RestController`를 사용했습니다.

### dto <-> domain 변환 로직 Mapper 구현
[관련 코멘트](https://github.com/woowacourse/spring-roomescape-admin/pull/200#discussion_r2050379924)

dusdn1702 commented on src/main/java/roomescape/controller/HomeController.java:
```diff
@@ -0,0 +1,12 @@
+public class HomeController {
+    @GetMapping("/")
+    public String redirectReservationPage() {
+        return "redirect:/admin/reservation";
```
https://github.com/woowacourse/spring-roomescape-admin/pull/200/files#r2050022585
추가해주신 부분 좋습니다. 다만, redirect 를 사용하신 이유가 있으신가요?

dusdn1702 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public boolean sameId(Long id) {
+        return this.id.equals(id);
```
여기서 this 를 추가한 이유와 위에는 추가하지 않으신 이유가 있으실까요? 궁금해요!

dusdn1702 reviewed (approved):
안녕하세요 멍구
피드백 잘 반영해주신 부분들 확인했어요.
고생 많으셨습니다.
남겨둔 몇가지 피드백은 다음 단계에서 반영해주셔도 될 것 같아, 이번 단계는 여기서 마무리 하도록 할게요. 
추가로 궁금하신 점 편하게 말씀 주시면 확인하도록 할게요. 😄
